client/src/lib.rs

```diff
@@ -306,6 +422,57 @@ impl<C> ConfiguredClient<C> {
     pub fn workers(&self) -> Arc<SlotManager> {
         self.workers.clone()
     }
+
+    /// Patches the client options, and then dynamically updates the endpoint with them.
+    /// If the update fails, the client does not rollback to the previous configuration, and future
+    /// connections will fail.
```

Comment:
This is rough. Why can't we only apply the update after the new endpoint is connected and works? Can't you do an atomic swap of the endpoint where there is no downtime and it only contains a successfully connected endpoint?

Reply:
Agreed. I would call this `try_update_options` and only swap if the connection succeeds. Also I find leaving the old options to be scary since now things aren't in sync.

It'd be fine to put a mutex around the options as well, so that they can be updated too. This also fits better with the idea of using the `ClientOptions` directly rather than `ClientOptionsUpdate`

Reply:
I've been looking into the atomic swap reusing the id for the endpoint, but the problem is that it s anything but atomic.
 When you reuse the id the endpoint gets in a "preparing" queue, the old endpoint still active, when the new one gets connected the swap happens. The problem is that this is all asynchronous, and if there is a problem with the new endpoint there is no visible error, and the old one continuous forever. It is also impossible to know when this swap will happen, or validate that it happened without sending a carefully crafted request. The endpoints are also created in lazy mode, so they need to be part of channel to validate them.
I think the root problem is that this mechanism was designed for load balancing, not dynamic changes, and if you are adding more (similar) endpoints with different targets, you can be more relaxed about when exactly changes happened.
For this reason, I'm forcing the client to fail when there are problems, otherwise they get silently ignored...

Reply:
> The problem is that this is all asynchronous

Yeah I see that with the balance channel and makes sense. But can we wait to even send the insert until we have connected the endpoint? And can we send the remove after the insert? So logic would be:

```
endpoint = connect_new()
channel_tx.send(insert(endpoint))
channel_tx.send(remove(old_endpoint_seq))
```

Also curious, if you send an insert then a remove, is it guaranteed that the _next_ call will be on the inserted?

Reply:
I don't think that you can easily swap the channel without creating a new client, and the problem of creating new clients is that there is a lot of shared state associated to them, like eager workflow start workers, or metrics,... that need to be moved across atomically.
The semantics of the control channel are really weak. Remove happens immediately, but inserts are queued in a "preparing" queue, so there is no guarantee that after sending insert ok you are getting the new options.
To avoid the gap you can reuse the id in insert, and then you don't need a remove, but there is no guarantee when, or if, the swap will happen.

Reply:
Thanks for the detail!

> I don't think that you can easily swap the channel without creating a new client, and the problem of creating new clients is that there is a lot of shared state associated to them, like eager workflow start workers, or metrics,... that need to be moved across atomically.

Yeah I see now that it gets a bit tangled there. I definitely wouldn't want a new top-level client, but I wonder if reasonable to put `TemporalServiceClient` or some of those even lower clients behind a mutex. We of course could do this all the way out on lang side if it wasn't for the need to change a workers' client. Currently this is what our customers do to switch clients, it's just workers that are the problems.

> Remove happens immediately, but inserts are queued in a "preparing" queue

So with the current code there's a potential to have no endpoints (i.e. remove went throw and insert is preparing) or does it read from the stream in that case?

Reply:
Yes, I had this comment in update_options
        // This approach may leave a short window in which the client cannot open new connections. Again,
        // we let the upper layer handle this infrequent failure.
I'm hoping that the retry would take care of it. In practice, the move from "preparing" queue to "ready" seems to be really quick, or triggered by the first request, and I have not been able to see it in the system tests. And changing the endpoint should only affect new requests, so the likelihood is much smaller.
But if all else fails, you could ended up with a no-endpoint client for a short while, and the error will propagate...

About doing surgery on the  low level client, not sure if there is some state that needs to be moved too, @Sushisource  can you comment?

Reply:
Swapping out `TemporalServiceClient`s would be reasonable, that's just a set of the lowest-level tonic clients. And, in fact, you don't even need a lock necessarily. You could probably use https://docs.rs/arc-swap/latest/arc_swap/ since it only needs to affect new requests.

However, the big problem there is clients wrap other clients generically using traits, so you don't know when you're down to the lowest level where you should be doing the swap. Changing everything to make that work would probably be annoying.

What Chad suggested w/ trying to connect the endpoint, and then sending it down the control channel if it works is what I was imagining. You can sort of arbitrarily make the whole thing sync by doing that, removing the old endpoint, adding the new one, then trying to call some rpc like server capabilities with some tight retry until it either works and you're done, or fails and you restore the old endpoint.

Reply:
> // This approach may leave a short window in which the client cannot open new connections. Again,
> // we let the upper layer handle this infrequent failure.

While our worker stuff is often self-recovering, the client stuff is not and some users use low-level pieces directly without explicit retry. I think we're going to have a hard time explaining a series of client failures during endpoint swap. We'll basically tell them to use their own swap if they want zero downtime because ours doesn't have that (and ignore worker warnings if we even log them). If at all possible, would like to avoid downtime just to change out the underlying client.

> However, the big problem there is clients wrap other clients generically using traits, so you don't know when you're down to the lowest level where you should be doing the swap

My naive guess would be to make every trait support updating options and have wrappers' impls call the underlying client until it reaches the true final client.

Reply:
>What Chad suggested w/ trying to connect the endpoint, and then sending it down the control channel if it works is >what I was imagining. You can sort of arbitrarily make the whole thing sync by doing that, removing the old endpoint, >adding the new one, then trying to call some rpc like server capabilities with some tight retry until it either works and >you're done, or fails and you restore the old endpoint.
Yes, I could make it a bit more robust by adding an rpc server check, like passing the ConfiguredClient and calling `get_system_info`, and try to undo the endpoint if it fails. However, other concurrent users of that client could still see a failure during that process, since there is no top level mutex, maybe that it fails for **everybody** gives us better error semantics. 
Trying to test the endpoint beforehand has less value. The api expects an endpoint in lazy mode, and the internal checks for moving from "preparing" to "ready" are repeated anyway. I will probably need to clone the endpoint and use a separate client to test it, and not sure we are gaining much over the "end-to-end" test of the rpc server check above...

Reply:
>While our worker stuff is often self-recovering, the client stuff is not and some users use low-level pieces directly without >explicit retry. I think we're going to have a hard time explaining a series of client failures during endpoint swap.

Do they remove the RetryClient part they get from the bridge, or you are thinking other SDKs like haskell?

I think the biggest value of this is for the worker, clients are easier to update with no need for dynamic stuff. 

But, as I said, the gap is theoretical, have not seen it in system tests that immediately try to use the updated client, and it may be that the request is on hold until the move from "preparing" to  "ready" happens, or is a quick local test...

Reply:
> Do they remove the RetryClient part they get from the bridge, or you are thinking other SDKs like haskell?

They make raw gRPC calls to the workflow service.

> But, as I said, the gap is theoretical

At the risk of over-simplifying, can we just do a zero-downtime client swap after guaranteed connection? Ideally this would work for all clients given new client options (or fail trying), but even if it just means a worker getting a `set_client`, that should be enough for most users and they will manage swapping their client uses their own ways.

Reply:
I think multiple endpoints is something else, we could do another PR for that.

Guaranteed connection before we start swapping, and get-system-info check before returning from update sounds good, will make things a bit more robust.

Now, zero-downtime is a trade-off vs predictability on when the change happens. I think this won't be a practical issue, specially when clients retry, and validation is quick, but if we don't delete the previous connection first, we cannot know when the new one is active, so get-system-info doesn't help. I frankly prefer losing a bit of availability to get better consistency.
 
Note that swapping the channel is actually worse for zero-downtime. It would affect existing connections, not only new ones, because returned values from the server may get ignored.

Given your feedback, my preference now is check before, check after, and tell people to retry for raw grpc, or use our RetryClient, or assume that once in a blue moon, you may get a connection error when upgrading...

Things that I am still on the fence:
  -Shall we try to restore the original endpoint if get-system-info fails? The error may already had caused other  users of this client failing. It may be better to have a clean, consistent error rather than a "some fail" type of error. And we check the connection before, so this would be a very rare event...
- We have `options()` and `options_mut()`. If we protect ClientOptions with a mutex, I'm thinking on returning a clone in `options()` and the Arc<Mutex> with `options_mut()`, so that `options()` may be inconsistent but won't block updates.
- With the mutex on ClientOptions get rid of ChangeSet and just update options directly internally.

Any thoughts? @Sushisource @cretz Shall I go ahead with the changes?

Reply:
> if we don't delete the previous connection first, we cannot know when the new one is active, so get-system-info doesn't help. I frankly prefer losing a bit of availability to get better consistency.

If this is behind a mutex, can we just assume that all outstanding calls may use old clients but all future calls will use new client? In other words, can all client use obtain the client from the mutex _for every gRPC request_ as needed (e.g. for worker polling) which means if the client is swapped it will be used for the next poll call but won't affect the current?

Reply:
Unfortunately the mutex is not top level, protecting the client, it is just guarding the options field (and  another one internal for update). Trying to change that now will break everything (i.e., all core SDKs bridges, cli,..)... And there are questions on what performance impact, or deadlocks, we will introduce by grabbing a lock, cloning,  on every request. 
 So my goal was to be as little intrusive as I could, even if things are not perfect...

Reply:
> Trying to change that now will break everything (i.e., all core SDKs bridges, cli,..)...

It shouldn't break anything, this should be an internal mutex to use the underlying thing. Maybe it is a channel and not a client, I don't know the word, but whatever the "thing that makes RPC calls is and closes itself on drop". If we call that `X`, I'd expect that `Mutex<Arc<X>>` would exist.

>  And there are questions on what performance impact, or deadlocks, we will introduce by grabbing a lock, cloning, on every request.

I don't think either is that much of a concern. The lock is only the life of an `Arc` clone I think. I think this is needed if you want to swap it and at negligible cost.

Reply:
Here's what ChatGPT suggested:

```rust
use std::sync::{Arc, RwLock};
use std::thread;

struct Client {
    // Client data
}

impl Client {
    fn new() -> Self {
        Client {
            // Initialization
        }
    }

    fn use_client(&self) {
        // Use the client
    }
}

fn main() {
    let client = Arc::new(RwLock::new(Client::new()));

    // Simulating a thread that uses the client
    let client_clone = Arc::clone(&client);
    thread::spawn(move || {
        let read_client = client_clone.read().unwrap();
        read_client.use_client();
    });

    // Simulating swapping the client
    let new_client = Client::new();
    let mut write_client = client.write().unwrap();
    *write_client = new_client;
}
```

Reply:
After chatting with ChatGPT, it ended up with `Arc<RwLock<Arc<Client>>` like so:

```rust
use std::sync::{Arc, RwLock};
use std::thread;

struct Client {
    // Client data
}

impl Client {
    fn new() -> Self {
        Client {
            // Initialization
        }
    }

    fn use_client(&self) {
        // Use the client
    }
}

fn main() {
    let client = Arc::new(RwLock::new(Arc::new(Client::new())));

    // Simulating a thread that uses the client
    let client_clone = Arc::clone(&client);
    thread::spawn(move || {
        let client_arc = client_clone.read().unwrap().clone();
        // The lock is released here, after cloning the Arc
        client_arc.use_client(); // use_client is called outside of the lock
    });

    // Simulating swapping the client
    let new_client = Arc::new(Client::new());
    let mut write_client = client.write().unwrap();
    *write_client = new_client;
}
```

Reply:
If we call `use_client` outside of the lock it doesn't help, right? We need to initiate the request holding the lock...

Interesting that chatGPT gave it a try, scary the number of bugs that it will create with subtle concurrency code...

Reply:
> If we call use_client outside of the lock it doesn't help, right? We need to initiate the request holding the lock...

I think it will help. Because that client is still under `Arc` so it is not dropped, it is just no longer on the field once swapped. So if you did a swap, once that `use_client()` call was done, it'd be the last reference and would drop it, but all others accessing the client get the new one.

To me, and again forgive my Rust rustiness, it is a case of atomically getting or swapping the client, but users of the client should not hold onto the lock while it is used. So you can swap a client out on the field that is in use by some function.

Reply:
Here's how it would look in Go if it reads better:

```go
type ClientHolder struct {
  // Client is thread safe
  myClient *Client
  myClientLock sync.RWLock
}

func (c *ClientHolder) Swap(newClient *Client) {
  c.myClientLock.Lock()
  defer c.myClientLock.Unlock()
  c.myClient = newClient
}

func (c *ClientHolder) Get() *Client {
  c.myClientLock.RLock()
  defer c.myClientLock.RUnlock()
  return c.myClient
}
```

Basically the `Get` client can be used by the caller regardless of whether it's still the _current_ client on the outer structure. We just expect the caller to use it for one request (again, it may no longer be the _current_ one) and call `Get` again for their next client need.

Reply:
The problem is that it is "lazy". It will create the connection with the first request, and we want to make sure that this connection is not created until we finish the upgrade.

Reply:
You should attempt a get-system-info call on the client before swapping it. IMO we want to make sure the connection _is_ created and is a known config to be successful even way before we swap it out. We don't want to swap out lazy or a when given a bad client, the worker will just silently fail for a minute (our default retry iirc) then fatal out.

Reply:
Before the swap we can check the connection, but that doesn't help for lazy. Internally Tonic defers the connection until it gets the first request. And even if we do a get-system-info, there is a race with someone else doing the first call after dropping the lock, if they grabbed it before the update started, and getting the error...

Reply:
I would probably do this:

```rust
pub struct ConfiguredClient<C> {
    client_and_options: ArcSwap<ClientAndOptions>,
    headers: Arc<RwLock<HashMap<String, String>>>,
    /// Capabilities as read from the `get_system_info` RPC call made on client connection
    capabilities: Option<get_system_info_response::Capabilities>,
    workers: Arc<SlotManager>,
}

pub struct ClientAndOptions<C> {
    client: C,
    options: ClientOptions
}

```

If ArcSwap doesn't work out for whatever reason (it might be hard to use `load()` references in all cases) then you can do the `Arc<RwLock<Arc<ClientAndOptions>>>` approach